Add Machine Type Support

realjenius opened the pull request:
## What does this PR do?

This PR adds optional support for the `MachineType` property in the Vultr provider spec, which enables creating a `BareMetalServer` instead of a cloud instance from Vultr.

This is based on reference from multiple stale branches from the AtomicJar fork where we implemented our own Vultr support... the most recent one being here: https://github.com/kubermatic/machine-controller/compare/main...AtomicJar:machine-controller:implement-vultr-cloud-provider

At this time, this PR is internal to the `aj-main` root branch, which is the forked publishing branch and is for internal testing. A cherry-picked version will be created for kubermatic at a later date.

## Why is it important?

We use bare metal servers with AJ TCC deployments, and this will enable us to use them with the upstream Vultr support, instead of our own forked version which is known to have bugs.

## Remaining Tasks
- [ ] Testing

realjenius committed:
Preliminary attempt to add support for machine types

realjenius committed:
Test data

realjenius committed:
Finish impl

lanwen commented on pkg/cloudprovider/provider/vultr/provider.go:
```diff
@@ -155,28 +156,48 @@ func (p *provider) Validate(ctx context.Context, _ *zap.SugaredLogger, spec clus
+	} else {
+		return fmt.Errorf("unknown machine type %q", c.MachineType)
 	}
-	if !planFound {
-		return fmt.Errorf("invalid/not supported plan specified %q: %w", c.Plan, err)
-	}
+
 	return nil
```
I think the latest `return nil` is unreachable and you can just return error from the `else` directly.

realjenius replied:
It actually turns out to be a catch all if none of the branches found an error, so it's reachable for the "everything is fine everywhere" case.

lanwen reviewed (approved):
I didn't get the origin of the change - is that the code from upstream just cherry-picked to include into our fork or it's our code? 

Would you be willing to schedule a pairing session to review this together? Meanwhile I'm good to merge it and try real-world, so we could pair whenever is convenient time for both

realjenius commented:
> I didn't get the origin of the change - is that the code from upstream just cherry-picked to include into our fork or it's our code?
> 
> Would you be willing to schedule a pairing session to review this together? Meanwhile I'm good to merge it and try real-world, so we could pair whenever is convenient time for both

Yes! The history is weird here, and it wasn't super well documented for me either. The short summary as I understand it is:

1. We added vultr support to machine controller in our fork (specifically Jojo did it)
2. There were several iterations on that across multiple branches
3. Alex consolidated all that work and did an upstream PR proposal - it was never accepted due to a lack of testing infrastructure for their e2e stack
4. Some time later, someone added a "community providers" concept which are providers that don't have E2E testing
5. Someone else then came along and added Vultr support again upstream - looks a lot like ours, but not the same. Notably it is missing MachineType support
6. This PR is my attempt at porting our support for MachineTypes on to the upstream Vultr support

[The other PR still in draft here](https://github.com/AtomicJar/machine-controller/pull/3) is a direct consequence of the outage, where we saw two Vultr servers with the same tags (i.e. assigned to the same Machine on the K8s side), which implies some sort of eventual consistency issue with their API and tags. So the idea instead is to switch to tracking the Vultr machine ID (since vultr has one) in k8s rather than tracking the Machine UID in Vultr.
